sound/soc/sof/pm.c

```diff
@@ -433,12 +433,48 @@ EXPORT_SYMBOL(snd_sof_set_d0_substate);
 
 int snd_sof_resume(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	int ret;
+
+	if (sdev->s0_suspend && sof_ops(sdev)->set_power_state) {
+		/* resume from D0I3 */
+		dev_dbg(sdev->dev, "ADSP will exit from D0i3...\n");
+		ret = snd_sof_set_d0_substate(sdev, SOF_DSP_D0I0);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed to exit from D0I3 %d\n",
+				ret);
+			return ret;
+		}
+
+		/* platform-specific resume from S0ix */
+		return snd_sof_dsp_resume(sdev);
+	}
+
+	/* resume from D3 */
 	return sof_resume(dev, false);
 }
 EXPORT_SYMBOL(snd_sof_resume);
 
 int snd_sof_suspend(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	int ret;
+
+	if (sdev->s0_suspend && sof_ops(sdev)->set_power_state) {
```

Comment:
I don't recall why we need to test for .set_power_state() being non-NULL. what does this have to do with the rest of the code in this patch?

Reply:
my intention here is that if no .set_power_state() ops defined, we will treat the corresponding platform has no D0I3 capability, for these platforms, we will run with D3 suspend/resume only.

Reply:
snd_sof_set_d0_substate() should be able to handle that case isnt it @keyonjie

Reply:
yeah, I don't understand why we are testing the same thing twice.

Reply:
um, this is a good point @ranj063 @plbossart , so here there are 2 problems:
1. To what platforms we should try to put to D013?
2. How should we handle the error cases for set_d0_substate()?

What you suggested here is combining the #1 into #2, but that will make the error handling more complex. Will add more comments about this later.

Reply:
Then add an abstraction such as 

````c
static inline bool is_d0_substate_supported(struct snd_sof_dev *sdev)
{
		if (sof_ops(sdev)->set_power_state)
                            return true;
                return false;
}
````

it's just ugly to use ops internals in generic code.